File: Server/Resources/Models/Product.cs
using System.Collections.Generic;
using Server.Resources.Enums;
using Server.Resources.Models.Core;

namespace Server.Resources.Models
{
    public class Product : Profile<Product>
    {
        public ProductKey Key { get; set; }
        public CompanyKey? CompanyKey { get; set; }
        public virtual Company Company { get; set; }
        public virtual List<Software> Software { get; set; }
        public virtual List<Service> Services { get; set; }

        public Product()
        {
            this.Software = new List<Software>();
            this.Services = new List<Service>();
        }
    }
}

File: Client/src/common/models/product.ts
import { Profile } from './core/profile';
import { Company } from './company';
import { Software } from './software';
import { Service } from './service';

export class Product extends Profile<Product> {
  public companyId?: number;
  public company?: Company;
  public software?: Array<Software>;
  public services?: Array<Service>;
}
